modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/LocalFileConfigurationStorage.java

```diff
@@ -0,0 +1,241 @@
+import com.typesafe.config.ConfigFactory;
+import com.typesafe.config.ConfigParseOptions;
+import com.typesafe.config.ConfigRenderOptions;
+import java.io.File;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+import org.apache.ignite.configuration.annotation.ConfigurationType;
+import org.apache.ignite.internal.configuration.NodeBootstrapConfiguration;
+import org.apache.ignite.internal.configuration.NodeConfigCreateException;
+import org.apache.ignite.internal.configuration.NodeConfigWriteException;
+import org.apache.ignite.internal.future.InFlightFutures;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteUtils;
+
+/**
+ * Implementation of {@link ConfigurationStorage} based on local file configuration storage.
+ */
+public class LocalFileConfigurationStorage implements ConfigurationStorage {
+    private static final IgniteLogger LOG = Loggers.forClass(LocalFileConfigurationStorage.class);
+
+    /**
+     * Path to config file.
+     */
+    private final Path configPath;
+
+    /**
+     * Path to temporary configuration storage.
+     */
+    private final Path tempConfigPath;
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    /**
+     * Latest state of last applied configuration.
+     */
+    private final Map<String, Serializable> latest = new ConcurrentHashMap<>();
+
+    /**
+     *  Configuration changes listener.
+     *  */
+    private final AtomicReference<ConfigurationStorageListener> lsnrRef = new AtomicReference<>();
+
+    private final ExecutorService threadPool = Executors.newFixedThreadPool(4, new NamedThreadFactory("loc-cfg-file", LOG));
+
+    private final InFlightFutures futureTracker = new InFlightFutures();
+
+    private long lastRevision = 0L;
+
+    /**
+     * Constructor.
+     *
+     * @param configuration Node bootstrap configuration.
+     */
+    public LocalFileConfigurationStorage(NodeBootstrapConfiguration configuration) {
+        this.configPath = configuration.configPath();
+        tempConfigPath = new File(configPath.toFile().getAbsolutePath() + ".storage").toPath();
```

Comment:
```suggestion
        tempConfigPath = configPath.resolveSibling(configPath.getFileName() + ".storage");
```

Reply:
I think this should work. You don't need to use `File` when you have `Path`, it's a legacy API